runtime/vm/BytecodeInterpreter.hpp

```diff
@@ -1186,10 +1186,23 @@ class INTERPRETER_CLASS
 					_currentThread->tempSlot = (UDATA)invalidIndex;
 					rc = THROW_AIOB;
 				} else {
-					I_32 value = VM_ArrayCopyHelpers::referenceArrayCopy(_currentThread, srcObject, srcStart, destObject, destStart, elementCount);
-					if (-1 != value) {
-						buildInternalNativeStackFrame(REGISTER_ARGS);
-						rc = THROW_ARRAY_STORE;
+					if (elementCount > 0) {
+						J9Class *srcComponentClass = ((J9ArrayClass *)J9OBJECT_CLAZZ(_currentThread, srcObject))->componentType;
+						J9Class *destComponentClass = ((J9ArrayClass *)J9OBJECT_CLAZZ(_currentThread, destObject))->componentType;
+
+						if (J9_IS_J9CLASS_PRIMITIVE_VALUETYPE(destComponentClass) &&
+							!J9_IS_J9CLASS_PRIMITIVE_VALUETYPE(srcComponentClass) &&
+							(NULL == J9JAVAARRAYOFOBJECT_LOAD(_currentThread, srcObject, srcStart))) {
```

Comment:
I don't see the check of flatten vs not flatten on the `srcComponentClass`. Does the NPE apply to the flattened case ?

Reply:
Only the first element being copied needs to be checked, because if the first element is both non-null and nullable then an ASE will be thrown instead

Reply:
Oh wait actually you're right, I didn't take into account the fact that primitive VTs could be inside an array of `Object`s. If you try to copy something like `new Object[] {new VTInt(), null, null}` into an array of `VTInt`s then you will still get a NPE, whereas `new Object[] {new Object(), null, null}` will get you an ASE. 

I can change the code to just check if the element at the index returned by `referenceArrayCopy` is null (since `referenceArrayCopy` returns the index at which an error occurred)

Reply:
> check if the element at the index returned by referenceArrayCopy is null

This should work.

Reply:
Ok I fixed this issue in [92cb616](https://github.com/eclipse-openj9/openj9/commit/92cb6169f6ef83030ce3d192917cf7b963164ba2)

I also added a check to make sure the source array is not flattened